### DISCUSSION TITLE
Outage 2023-02-17?

### DISCUSSION THREAD

**Discussion by richcorbs** [pockethost/pockethost#185] [Q&A]
Are we having an outage?

**Comment by MalikWhitten67** [comment]
yeah,  - postr went unresponsive for a bit thought I sent too many requests. 😅.  It was off and on, highest request time was 3000ms then it blacked out again.

**Comment by benallfree** [comment]
Hi everyone, there is indeed an outage. Looking into it now.

**Comment by benallfree** [comment]
Ok, we should be back up. Thanks for the notification. I'm grateful to all of you who continue dogfooding this service with me :)

**Reply by MalikWhitten67** [reply]
Lol no worries

**Comment by bloor** [comment]
Today/Now it's down as well. 
Very sadge. I use weekend to code and expand my skillz on nuxt and pocketbase. 
Guess I go hunt some food and hopefully it recovers.

**Reply by Mellbin** [reply]
@benallfree any ideas on what's happening?

**Reply by bloor** [reply]
Could be someone doing some heavy stress-tests on his db instance, but in turn overloading the main domain since the homepage also not working. Main server at limit.

**Reply by Mellbin** [reply]
We will see, but if that's true maybe a good idea to implement some kind of roof for raw usage per instance.
Although we should not make demands on "SLA's" since this is free and super generous of Ben it would be nice with some stability of the platform.
It seems very easy to create malicious software to overload the servers, on purpose or not.

**Reply by benallfree** [reply]
We are back up again. Looking through logs now.

**Comment by benallfree** [comment]
Hi everyone, just a note as I continue to investigate the cause of the recent outages. These things are bound to happen in production where they don't show up in development. 

In both outages, I found one or more PocketBase processes eating up all the CPU. This is surprising because it means PocketBase itself is doing a lot of processing work and not stopping, which I have not seen before. @ganigeorgiev you might be interested in collaborating to understand what is going on :) 

```
 PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                                      
 413028 root      20   0  763328  32636  23264 S  88.4   0.8  15:34.48 pocketbase                                                                                                                                                                   
 390278 root      20   0   31.0g 243000  41076 R  57.5   6.0  72:21.57 node                                                                                                                                                                         
 390316 root      20   0   20.9g 124152  37076 R  37.2   3.1  38:09.41 node                                                                                                                                                                         
 390471 root      20   0  639768  89856  33492 R   8.6   2.2   6:28.13 node                                                                                                                                                                         
```

Perhaps we need some way to detect whether a proxied request to PocketBase has taken too long and then kill the PocketBase process? I'm not sure what to do in cases where PocketBase is misbehaving.

**Reply by bloor** [reply]
Not that hard to bottleneck any SQL db. If u spam reads/updates on M+ rows and then loop it in a script that runs each minute. 

Or u simply make a chat app and dump 10k+ users to stress test it.

In both cases seems like stress test scenario.
Reading those logs, given high CPU and low RAM usage seeems like simple query spam rather than long complex query processes.

**Reply by benallfree** [reply]
@bloor Maybe. The ability to send any kind of query that overloads PocketBase seems dangerous. PocketHost is mostly just acting as a proxy to PocketBase, so if PocketBase is allowing something to lock it up, that's not good for anyone. That's why I'm curious to hear from @ganigeorgiev to see if this is something he already knows about.

**Reply by bloor** [reply]
Well the (web)user can be restricted to send queries, but the admin (the owner of the pocketbase can do any query as complex as he sees fits). Pocketbase itself is in beta, and on top of that is advertised as a VPS/self-host solution. 

In a normal scenario where I host Pocketbase on my own server, I get capped by the server capabilities and I can push it as hard as I can since I`m the only one using the server. Any similar pocketbase-host solution like you are trying to offfer, that uses 1 server to host many instances , requires to implement a instance-cap since the CPU/RAM is shared.

This is also probably why the Pocketbase author is not foreseeing to provide any hosting solution free/paid anytime soon. It is a hurdle to setup the server properly.

IDK, the way I see things, since you worked hard and offered a free solution, for time being you should do some log-flagging and warnings/triggers, and any account which goes overboard on resources be warned/suspended etc.

**Reply by benallfree** [reply]
Another thought... Maybe there is a way to launch each PocketBase instance in its own virtual machine rather than a shared environment. I know that fly.io has Fly Machines https://fly.io/blog/fly-machines/ and google has Cloud Run... maybe Digital Ocean has something similar.

**Reply by benallfree** [reply]
@bloor imho PocketBase should not allow *any* query (admin or not) to run endlessly. That's my question for @ganigeorgiev. And, I'm a little unclear on why the Linux process scheduler isn't properly multitasking. The problem could also be in the PocketHost code somewhere (likely!). Perhaps the `node` process is becoming unresponsive and that's why it presents as an outage. 

But either way, I suppose PocketHost should handle situations where PocketBase is misbehaving. It wouldn't be hard to add a timeout to simply kill a PocketBase process that has been taking too long to return query results. But this could cause thrashing in a multitenant environment, so I have to think about how to find the right PocketBase process to kill :)

**Reply by ganigeorgiev** [reply]
@benallfree It is hard to say what exactly is causing the issue. PocketBase doesn't really have very CPU intensive operations, but I guess high CPU usage could be caused by:

- **large number of sign-ins/sign-ups** (this is because generating the bcrypt hash and verifying it could be slow; this could also explain the high-cpu:low-memory usage in the above log)

- high requests load (both read and writes)

- complex queries across large datasets (this include complex rules with `@collection.*` joins)

- thumbs generation of very large images


> imho PocketBase should not allow any query (admin or not) to run endlessly. That's my question for @ganigeorgiev.

Currently PocketBase doesn't have a central query timeout handling. I'm planning implementing fail/retry for the select queries and will consider specifying a cancellable context in the next v0.13 or v0.14 release. But based on the above usage log I doubt that this is the case here because I'd expect also high memory consumption with the slow and complex queries.

**Reply by ganigeorgiev** [reply]
@benallfree  Just for info, the above high CPU usage could indeed be caused by a long running query. I was able to reproduce the issue locally by emulating multiple joins across large datasets. The issue could be also reproduced by running a recursive expression like:

```sql
WITH RECURSIVE r(i) AS (
  VALUES(0)
  UNION ALL
  SELECT i FROM r
  LIMIT 100000000
)
SELECT i FROM r WHERE i = 1;
```

I've added a default `SELECT` query timeout of 60s in the [`develop` branch](https://github.com/pocketbase/pocketbase/tree/develop) together with fail/retry handling for the "database is locked" `SELECT` errors during mixed read&write transactions.
The fixes will be available in the next v0.13.0 release (_sometime at the end of the week or the week following it_), but please keep in mind that users will still be able to abuse the shared free service (_and that will be even more easier with the ["view" collection type](https://github.com/pocketbase/pocketbase/pull/759#issuecomment-1435731789) in the upcoming release since **users will be able to execute any custom `SELECT` query**_), meaning that you still have to implement some additional guards to prevent this from happening - restrict a single PocketBase process to use the CPU above some threshold for max amount of time, soft-limit the memory usage (could be done also globally with the `GOMEMLIMIT` env variable), etc.

**Reply by benallfree** [reply]
@ganigeorgiev That sounds great! It might be nice if that timeout were configurable, for the free tier I'd probably reduce it to 5-10 seconds rather than 60.

Since you're doing stress testing, could you try [cpulimit](https://manpages.ubuntu.com/manpages/bionic/man1/cpulimit.1.html) to see if long-running queries can be CPU-limited by `SIGSTOP` and `SIGCONT` or other signals?

**Reply by ganigeorgiev** [reply]
@benallfree Currently the default timeout is configurable only when extending PocketBase (eg. `app.Dao().ModelQueryTimeout = 10 * time.Second`), but I'll consider defining a command line flag(s) for the prebuilt executable.

> Since you're doing stress testing, could you try [cpulimit](https://manpages.ubuntu.com/manpages/bionic/man1/cpulimit.1.html) to see if long-running queries can be CPU-limited by SIGSTOP and SIGCONT or other signals?

I've tested with `cpulimit -p 733245 -l 50` and it throttles the PocketBase process to max ~50% (_you may also want to explore the `-k` and `-m` flags_).

**Comment by benallfree** [comment]
I'm seeing intermittent CPU spiking again. I might try upgrading the server to see if that settles things down. Perhaps it's just getting busier. 

<img width="979" alt="image" src="https://user-images.githubusercontent.com/1068356/219875602-1a552146-7027-493e-94e9-83bd178aa6bc.png">

**Reply by benallfree** [reply]
Seems like the resize worked...

<img width="997" alt="image" src="https://user-images.githubusercontent.com/1068356/219877826-df0f4350-cd37-41b1-9fac-aa4e4eae2be4.png">

**Comment by benallfree** [comment]
Okay... resized to 16 GB / 8 vCPUs, let's see if that changes anything!

**Reply by benallfree** [reply]
PS... currently supporting over 1,200 instances. Historically, about 10% (120) are highly active. So we are definitely growing!

**Reply by Mellbin** [reply]
Would it be to difficult to setup multiple VM's?  To increase redundancy and "load-balance" incase something goes wrong with the other one?
etc xxx.pockethost.io and xxx.pockethost2.io? 

Not sure if it's cost efficient to have more less powerful vm's or not though.

**Reply by benallfree** [reply]
@Mellbin Yes, the next step to scaling would be to write some 'pockethost worker' code and spin up multiple workers. To do that, we'd need to see how SQLite handles over a networked file system and we'd need to make sure only one PocketBase runs per instance.

**Reply by bloor** [reply]
I`m reluctant that resizing server would do much if those leech-instances keep going 100% cpu. They will simply suck even more resources.

**Reply by bloor** [reply]
https://askubuntu.com/questions/1367612/how-can-i-limit-the-cpu-and-ram-usage-for-a-process

**Comment by richcorbs** [comment]
Is it happening again? Is it me causing this?

**Reply by bloor** [reply]
kind of.. my instance stopped loading right now.

**Reply by richcorbs** [reply]
Main site is down. My instance is down.

**Reply by bloor** [reply]
it comes and goes 👀
it;s on now .. CODE CODE CODE CODE !

**Reply by benallfree** [reply]
Hmm... well we could try [cpulimit](https://manpages.ubuntu.com/manpages/bionic/man1/cpulimit.1.html) to keep pocketbase PID's from taking over everything... potentially. Need to research more first, I'm not clear what is causing these spikes.

**Comment by richcorbs** [comment]
I'm available and willing most of the day today to help if I can.

**Reply by benallfree** [reply]
@richcorbs what are you doing that you think would be causing it?

**Reply by richcorbs** [reply]
Just using the system. I don't think I'm doing anything specifically that would trigger anything like this.

I'm trying to be a good citizen in my code. I have a couple of tables with thousands of rows and I've tried not to do many expands across them.

However, here is the one thing that I think *might* be related:

```
this.currentActivity = await this.pb.collection('activities').getOne(<id>, {
  expand: 'attendees,directors,locations,projects',
  sort: '+attendee.firstname'
})
```

**Reply by benallfree** [reply]
Hmm I think you should be able to run whatever query you want. I'm not clear on why the linux kernel is not multitasking more effectively. There aren't *that* many processes running at a given time. I'm not convinced that even a rogue pocketbase process should cause this kind of outage.

**Reply by richcorbs** [reply]
Clicking around now is snappy. Trying to exercise that bit of code.

If you think of anything I can do to help I'm willing.

**Comment by richcorbs** [comment]
Is it happening again?

**Reply by benallfree** [reply]
Yes it appears to be PocketBase.

<img width="623" alt="image" src="https://user-images.githubusercontent.com/1068356/219909045-4e7ff133-3b40-496b-897d-80c09ef07996.png">

@ganigeorgiev what do you think?

**Reply by benallfree** [reply]
All I can think to do is try `cpulimit`, or somehow move each PocketBase process to its own VPS.

**Reply by benallfree** [reply]
I bet I could write some type of provider that would launch PocketBase in a Dockerized container on a separate VPS. So there could be a Digital Ocean provider, a 'shared CPU' provider that launches on the same CPU, and we could write providers basically for every cloud platform. That would be a cool enhancement for PocketHost.

**Reply by richcorbs** [reply]
Can you connect a PID to an instance of pocketbase?

**Reply by benallfree** [reply]
Yes

**Reply by ganigeorgiev** [reply]
@benallfree I'm not really sure what is causing the above. At least based on my local tests, the live demo (aka. pocketbase.io) and its Hetzner CPU graph for the last 30 days, there is no unusual resources consumption and the CPU usage is usually below ~ 5-10% (_note that the live demo is on one of the cheapest instance with only 2vCPU_):
![cpu_graph](https://i.imgur.com/fxR9idp.png)

I suspect that someone is spamming your instance or running some heavy computation task but without logs I'm not really sure.

Are users allowed to execute custom code as part of a hook event?

Are there any excessive access logs or errors?

In addition to the `cpulimit`, if you haven't already, you can try to specify a rate limiter on a reverse proxy level (eg. [nginx-rate-limiting](https://www.nginx.com/blog/rate-limiting-nginx/)) and restrict the max upload body size globally (eg. with `client_max_body_size` for nginx).

**Reply by benallfree** [reply]
@ganigeorgiev I'm not seeing any excessive error logs unfortunately. This one is hard to track down though, I'd need to add more logging to see specifically what query is being proxied that is causing such severe spikes. 

No custom code is allowed.

Good idea about max upload and API rate limiting.

Another thought I had was to spin up each instance in its own VM using a Docker image. Are you shipping docker images with each version release yet? If not, that might be an interesting addition to your release automation. 

edit: to expand on the VM approach a little, I would create a storage volume for each instance and then mount that volume to a VM when the instance started. Then I'd shut down the VM during inactivity. I think this is somewhat similar to what Supabase does.

**Reply by ganigeorgiev** [reply]
@benallfree 

>  Are you shipping docker images with each version release yet?

No, at the moment there is no official Docker image yet. For a simple Dockerfile you can check as a reference https://pocketbase.io/docs/going-to-production/#using-docker (_practically it just downloads and unzip the release archive from the github action assets /note that they have some rate limiter per request ip or access token/_).

The individual VM/container approach in general I think would be better for your use case (_although it may complicate the networking configuration and will take a little more space_), since it will ensure instances isolation and  will be easier to track and apply usage limits per user/instance.

**Reply by bloor** [reply]
😭5h+ downtime

**Reply by Mellbin** [reply]
@bloor shit happens, I moved my DB to a Azure VM today ☁️ and damn my response times went down by 70-80% but then again it costs 💸 
I'll maybe go back when we are reaching a more stable release

**Reply by bloor** [reply]
Well, I could spin a 5$ vps as well, money not a problem. It's just that it takes time to configure and i`m a linux pleb.
I already have a project hosted on a VPS pocketbase instance, but it took me 2-3 days to properly configure ubuntu.
I came here for POC and fast spin up projects. I'm still testing pocketbase capabilities.

Pokethost.io gives some ease of mind, when it works.

**Reply by Mellbin** [reply]
agreed, took me 4 hours to get everything running with DNS, auto start and all, now I could prob do it under 1h but yeah if/when PocketHost is running at 1.0.0 it will be great and so many weekend projects will be able to use it.
To be clear Im not abandoning this project, I just need some stability as I plan to release my app to users next week.

**Reply by benallfree** [reply]
@bloor @Mellbin Downtime is super frustrating! On the other hand, it means PH is growing and being used, and these are the types of issues that are hard to see in a controlled development environment. I think isolating the PB instances is the only way we can really stop this from happening. 

In the mean time, I've set up something that will restart all the instances every 15 minutes. That should keep things from being down for too long until I can solve this in a more permanent way.

**Reply by bloor** [reply]
> agreed, took me 4 hours to get everything running with DNS, auto start and all, now I could prob do it under 1h but yeah if/when PocketHost is running at 1.0.0 it will be great and so many weekend projects will be able to use it. To be clear Im not abandoning this project, I just need some stability as I plan to release my app to users next week.

Yas! PB barely at 0.12 and already offers 95% of my needs. I can't wait to see what it can offer at 1.0 release. Haven't been so excited for dev stuff since Vue2.
